File: src/main/java/tlint/CLIRunner.java
package tlint;

import com.google.common.base.Charsets;
import com.intellij.execution.ExecutionException;
import com.intellij.execution.configurations.GeneralCommandLine;
import com.intellij.execution.process.*;
import com.intellij.openapi.util.Key;
import org.jetbrains.annotations.NotNull;
import java.io.File;

public final class CLIRunner {
    /**
     * @param cwd working directory
     * @param node node interpreter path
     * @param exe node executable to run
     * @return command line to execute
     */
    @NotNull
    public static GeneralCommandLine createCommandLine(@NotNull String cwd, String filePath, String tlintPath) {
        GeneralCommandLine commandLine = new GeneralCommandLine();

        if (!new File(cwd).exists()) {
            throw new IllegalArgumentException("cwd doesn't exist");
        }
        commandLine.setWorkDirectory(cwd);

        if (!new File(cwd + "/" + filePath).exists()) {
            throw new IllegalArgumentException("filePath doesn't exist");
        }

        if (!new File(tlintPath).exists()) {
            throw new IllegalArgumentException("tlint path doesn't exist");
        }
        commandLine.setExePath(tlintPath);

        return commandLine;
    }

    @NotNull
    public static ProcessOutput execute(@NotNull GeneralCommandLine commandLine, int timeoutInMilliseconds) throws ExecutionException {
        Process process = commandLine.createProcess();
        OSProcessHandler processHandler = new ColoredProcessHandler(process, commandLine.getCommandLineString(), Charsets.UTF_8);
        final ProcessOutput output = new ProcessOutput();

        processHandler.addProcessListener(new ProcessAdapter() {
            public void onTextAvailable(ProcessEvent event, Key outputType) {
                if (outputType.equals(ProcessOutputTypes.STDERR)) {
                    output.appendStderr(event.getText());
                } else if (!outputType.equals(ProcessOutputTypes.SYSTEM)) {
                    output.appendStdout(event.getText());
                }
            }
        });

        processHandler.startNotify();

        if (processHandler.waitFor(timeoutInMilliseconds)) {
            output.setExitCode(process.exitValue());
        } else {
            processHandler.destroyProcess();
            output.setTimeout();
        }

        if (output.isTimeout()) {
            throw new ExecutionException("Command '" + commandLine.getCommandLineString() + "' is timed out.");
        }

        return output;
    }
}

File: src/main/kotlin/tlint/CLIRunner.kt
package tlint

import com.google.common.base.Charsets
import com.intellij.execution.ExecutionException
import com.intellij.execution.configurations.GeneralCommandLine
import com.intellij.execution.process.*
import com.intellij.openapi.util.Key
import java.io.File

object CLIRunner {
    fun createCommandLine(cwd: String, filePath: String, tlintPath: String): GeneralCommandLine {
        val commandLine = GeneralCommandLine()

        if (!File(cwd).exists()) {
            throw IllegalArgumentException("cwd doesn't exist")
        }
        commandLine.setWorkDirectory(cwd)

        if (!File("$cwd/$filePath").exists()) {
            throw IllegalArgumentException("filePath doesn't exist")
        }

        if (!File(tlintPath).exists()) {
            throw IllegalArgumentException("tlint path doesn't exist")
        }
        commandLine.exePath = tlintPath

        return commandLine
    }

    @Throws(ExecutionException::class)
    fun execute(commandLine: GeneralCommandLine, timeoutInMilliseconds: Int): ProcessOutput {
        val process = commandLine.createProcess()
        val processHandler = ColoredProcessHandler(process, commandLine.commandLineString, Charsets.UTF_8)
        val output = ProcessOutput()

        processHandler.addProcessListener(object : ProcessAdapter() {
            override fun onTextAvailable(event: ProcessEvent, outputType: Key<*>) {
                if (outputType == ProcessOutputTypes.STDERR) {
                    output.appendStderr(event.text)
                } else if (outputType != ProcessOutputTypes.SYSTEM) {
                    output.appendStdout(event.text)
                }
            }
        })

        processHandler.startNotify()

        if (processHandler.waitFor(timeoutInMilliseconds.toLong())) {
            output.exitCode = process.exitValue()
        } else {
            processHandler.destroyProcess()
            output.setTimeout()
        }

        if (output.isTimeout) {
            throw ExecutionException("Command '" + commandLine.commandLineString + "' is timed out.")
        }

        return output
    }
}